### PR TITLE
Pass Analytics parameters to redirect URL

### PR DESCRIPTION
**Pull request by alistat**

### CODE REVIEW

File: src/analytics_url.coffee [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+    URL_PARAM_NAME = 'skr_info'
+    TTL = 60 * 1000 # milliseconds
+
+    constructor: (@url) ->
+
+    format_params: (mode, session, metadata = {}) ->
+      params = [session, @_expiration(), metadata]
+      param_hash = Base64.encodeURI JSON.stringify(params)
+
+      return switch mode.type
```
Comment (kosdiamantis) [inline src/analytics_url.coffee]: `return` is redundant.

File: src/analytics_url.coffee [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+  'helpers/base64_helper'
+], (URLHelper, Base64)->
+  class AnalyticsUrl
+    URL_PARAM_NAME = 'skr_info'
+    TTL = 60 * 1000 # milliseconds
+
+    constructor: (@url) ->
+
+    format_params: (mode, session, metadata = {}) ->
+      params = [session, @_expiration(), metadata]
```
Comment (kosdiamantis) [inline src/analytics_url.coffee]: Maybe we should use a js object here instead of Array. 🤔

What do you think?

File: src/actions_manager.coffee [hunk 3]
```diff
@@ -67,11 +68,6 @@ define [
         ## TODO: THROW ERROR ON REJECT
         @reporter.sendBeacon(url, payload).then => cb and cb(@session.analytics_session)
```
Comment (kosdiamantis) [inline src/actions_manager.coffee]: I suggest to return `@session` here: `cb(@session)` in order to have full access in session object in `productClick` action.

File: src/analytics_url.coffee (hunk 2 above)
Comment (alistat) [reply src/analytics_url.coffee]: I had it as a JS object but it was almost double the length..

File: src/actions_manager.coffee (hunk 3 above)
Comment (alistat) [reply src/actions_manager.coffee]: What if we don't pass anything to the callback so that we can use the value if we want to signify an error later?

Comment (alistat) [reply src/actions_manager.coffee]: It seemed a little arbitrary to return the session...

File: src/analytics_url.coffee [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+      try
+        params = JSON.parse(url_param)
+      catch err # invalid JSON
+        return null
+
+      [session, expiration, metadata] = params
+
+      {
+        session,
+        expired: expiration < Date.now()
```
Comment (kosdiamantis) [inline src/analytics_url.coffee]: I suggest to return null if it is expired.

File: spec/analytics_url_spec.coffee [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+  describe '#read_params', ->
+    beforeEach ->
+      @url = new @AnalyticsUrl('https://shop.gr/product').format_params(@mode, @session, @metadata)
+      @params = new @AnalyticsUrl(@url).read_params()
+
+    it 'reads session', ->
+      expect(@params.session).to.eq(@session)
+
+    it 'reads metadata', ->
+      expect(@params.metadata).to.eql(@metadata)
```
Comment (alistat) [inline spec/analytics_url_spec.coffee]: `.deep.equal` seems more explicit..

**Pull request merged**